### ISSUE TITLE
optional does not work well with custom

### ISSUE THREAD

**Issue by blacksteed232** [mobxjs/serializr#128]
Hi,

I tried to define a property schema nesting `custom` inside an `optional` for my serializer and it does not seem to work well. I have looked at the codebase and came up with the following test case:

```javascript
test("optional properties should work with `custom` serializer", t => {
    var schema = {
        factory: () => ({}),
        props: {
            optionalProp: optional(custom(
                value => value.id,
                () => SKIP
            )),
        }
    }
    t.deepEqual(serialize(schema, { optionalProp: undefined }), {})
    t.end()
})
```

This will throw `TypeError: Cannot read property 'id' of undefined`

My expectation is `optional` will first check if the value is `undefined`, if so it returns `SKIP`, otherwise it will proceed, am I missing something here?

**Comment by pyrogenic** [comment]
`optional` simply avoids serializing properties with `undefined` values. When wrapped with `optional`, if your custom serializer returns `undefined`, no property will appear in the JSON at all. If you want `undefined` to appear in the JSON, you don't need `optional` — just return `undefined` from your custom serializer.

**Comment by blacksteed232** [comment]
@pyrogenic then I guess `optional` returns `SKIP` if the value it wraps is `undefined`. Maybe we should clarify the doc a bit about this. Since my perception is `serializer` are read from outer to inner -- e.g. `list(object())` will do `list()` first, which led me to believe `optional(custom())` will do `optional()` check first, but it didn't

**Comment by pyrogenic** [comment]
> I guess optional returns SKIP if the value it wraps is undefined

That is [exactly correct](https://github.com/mobxjs/serializr/blob/master/src/types/optional.ts#L23-L16) 😄 
```js
    function serializer(...args) {
        const result = propSerializer(...args)
        if (result === undefined) {
            return SKIP
        }
        return result
    }
```

I think I understand your confusion — I wouldn't think of the `optional` decorator as a "check". Both `list` and `optional` operate on the _result_ of using the using the given schema to serialize data.

You're welcome to suggest a change to [the docs](https://github.com/mobxjs/serializr/blob/master/src/types/optional.ts#L5-L16).

```js
/**
 * Optional indicates that this model property shouldn't be serialized if it isn't present.
 *
 * @example
 * createModelSchema(Todo, {
 *     title: optional(primitive()),
 * });
 *
 * console.dir(serialize(new Todo()));
 * // {}
 *
 * @param {PropSchema} propSchema propSchema to (de)serialize the contents of this field
 * @returns {PropSchema}
 */
```

**Comment by blacksteed232** [comment]
@pyrogenic gotcha, I have submitted a tiny PR #130. Could you give review that? Thanks!

**Comment by pyrogenic** [comment]
Resolved with #131

**Comment by pyrogenic** [comment]
Published new docs @ https://www.npmjs.com/package/serializr/v/2.0.3
